modules/seqexec/server/src/main/scala/seqexec/server/Systems.scala

```diff
@@ -43,135 +47,286 @@ final case class Systems[F[_]](
-    if (settings.gmos.command) GmosSouthControllerEpics(GmosEpics.instance).pure[IO]
-    else                              GmosControllerSim.south[IO]
-
-  def gmosNorth(settings: SystemsControlConfiguration)(implicit T: Timer[IO], L: Logger[IO]): IO[GmosNorthController[IO]] =
-    if (settings.gmos.command) GmosNorthControllerEpics(GmosEpics.instance).pure[IO]
-    else                              GmosControllerSim.north[IO]
-
-  def flamingos2(settings: SeqexecEngineConfiguration)(implicit T: Timer[IO], L: Logger[IO]): IO[Flamingos2Controller[IO]] =
-    if (settings.systemControl.f2.command) Flamingos2ControllerEpics[IO](Flamingos2Epics.instance).pure[IO]
-    else if (settings.instForceError) Flamingos2ControllerSimBad[IO](settings.failAt)
-    else Flamingos2ControllerSim[IO]
-
-  def gpi[F[_]: ConcurrentEffect: Timer](settings: SeqexecEngineConfiguration, httpClient: Client[F]): Resource[F, GpiController[F]] = {
-    def gpiClient: Resource[F, GpiClient[F]] =
-      if (settings.systemControl.gpi.command) GpiClient.gpiClient[F](settings.gpiUrl.renderString, GpiStatusApply.statusesToMonitor)
-      else                             GpiClient.simulatedGpiClient[F]
-
-    def gpiGDS: Resource[F, GdsClient[F]] =
-      Resource.pure(GdsClient(
-        if (settings.systemControl.gpiGds.command) httpClient else GdsClient.alwaysOkClient[F],
-        settings.gpiGDS))
-
-    (gpiClient, gpiGDS).mapN(GpiController(_, _))
-  }
 
-  def ghost[F[_]: ConcurrentEffect: Timer](settings: SeqexecEngineConfiguration, httpClient: Client[F]): Resource[F, GhostController[F]] = {
-    def ghostClient: Resource[F, GhostClient[F]] =
-      if (settings.systemControl.ghost.command) GhostClient.ghostClient[F](settings.ghostUrl.renderString)
-       else                              GhostClient.simulatedGhostClient
+  final case class Builder(settings: SeqexecEngineConfiguration, service: CaService, tops: Map[String, String])(implicit L: Logger[IO]) {
+    def odbProxy[F[_] : Sync : Logger]: OdbProxy[F] = OdbProxy[F](new Peer(settings.odb.renderString, 8443, null),
+      if (settings.odbNotifications) OdbProxy.OdbCommandsImpl[F](new Peer(settings.odb.renderString, 8442, null))
+      else new OdbProxy.DummyOdbCommands[F])
 
-    def ghostGDS: Resource[F, GdsClient[F]] =
-      Resource.pure(GdsClient(
-        if (settings.systemControl.ghostGds.command) httpClient else GdsClient.alwaysOkClient[F],
-        settings.ghostGDS))
+    def dhs[F[_] : Effect : Timer : Logger](httpClient: Client[F]): F[DhsClient[F]] =
+      if (settings.systemControl.dhs.command)
+        DhsClientHttp[F](httpClient, settings.dhsServer).pure[F]
+      else
+        DhsClientSim.apply[F]
 
-    (ghostClient, ghostGDS).mapN(GhostController(_, _))
-  }
+    // TODO make instruments controllers generalized on F
+    def gcal(implicit L: Logger[IO]): IO[(GcalController[IO], GcalKeywordReader[IO])] =
```

Comment:
Given that `instance` is parametrized I'd expect we could make these parametric too. Why don't we?

Reply:
The 'F' in `instance` instance is the effect to produce the EPICS class. `EpicsSystem[T].instance[F]` produces a `F[T]`, but in all cases `T` is something like `TcsEpics[IO]`, `GmosEpics[IO]`, etc. We are not yet at the point where we could get rid of that IO, because the different `XXXEpics[IO]` are implemented with objects.